### ISSUE TITLE
RaymondFam - Normal users could be inadvertently grieved by the withdrawn ratios check

### ISSUE THREAD

**Issue by sherlock-admin** [sherlock-audit/2023-03-olympus-judging#28]
RaymondFam

high

# Normal users could be inadvertently grieved by the withdrawn ratios check

## Summary
The contract check on the withdrawn ratios of OHM and wstETH against the current oracle price could run into grieving naive users by taking any wstETH shifted imbalance as a fee to the treasury even though these users have not gamed the system. 

## Vulnerability Detail
Here is a typical scenario, assuming the pool has been initiated with total LP equal to sqrt(100_000 * 1_000) = 10_000. (Note: OHM: $15, wstETH: $1500 with the pool pricing match up with [manager.getOhmTknPrice()](https://github.com/sherlock-audit/2023-03-olympus/blob/main/sherlock-olympus/src/policies/BoostedLiquidity/BLVaultLido.sol#L156) or [manager.getTknOhmPrice()](https://github.com/sherlock-audit/2023-03-olympus/blob/main/sherlock-olympus/src/policies/BoostedLiquidity/BLVaultLido.sol#L232), i.e. 100 OHM to 1 wstETH or 0.01 wstETH to 1 OHM. The pool token balances in each step below may be calculated via the [Constant Product Simulation](https://amm-calculator.vercel.app/) after each swap and stake.)

    OHM token balance: 100_000
    wstETH token balance: 1_000
    Total LP: 10_000

1. A series of swap activities results in the pool shifted more of the LP into wstETH.

    OHM token balance: 90_909.1
    wstETH token balance: 1_100
    Total LP: 10_000

2. Bob calls [`deposit()`](https://github.com/sherlock-audit/2023-03-olympus/blob/main/sherlock-olympus/src/policies/BoostedLiquidity/BLVaultLido.sol#L143) by providing 11 wstETH where 1100 OHM is minted with 1100 - 90909.1 * 0.01 = 190.91 unused OHM [burned](https://github.com/sherlock-audit/2023-03-olympus/blob/main/sherlock-olympus/src/policies/BoostedLiquidity/BLVaultLido.sol#L187-L190). (Note: Bob successfully stakes with 909.09 OHM and 11 wstETH and proportionately receives 100 LP.)

    OHM token balance: 91_818.19
    wstETH token balance: 1_111
    Total LP: 10_100
    User's LP: 100

3. Bob changes his mind instantly and proceeds to call [`withdraw()`](https://github.com/sherlock-audit/2023-03-olympus/blob/main/sherlock-olympus/src/policies/BoostedLiquidity/BLVaultLido.sol#L203) to remove all of his LP. He receives the originally staked 909.09 OHM and 11 wstETH. All OHM is burned but he is only entitled to receive 909.09 / 100 = 9.09 wstETH since the system [takes any arbs](https://github.com/sherlock-audit/2023-03-olympus/blob/main/sherlock-olympus/src/policies/BoostedLiquidity/BLVaultLido.sol#L236-L240) relative to the oracle price for the Treasury and [returns the rest](https://github.com/sherlock-audit/2023-03-olympus/blob/main/sherlock-olympus/src/policies/BoostedLiquidity/BLVaultLido.sol#L247) to the owner.

    OHM token balance: 90_909.1
    wstETH token balance: 1_100
    Total LP: 10_000
    User's LP: 0

## Impact
Bob suffers a loss of 11 - 9.09 = 1.91 wstETH  (~ 17.36% loss), and the system is ready to trap the next user given the currently imbalanced pool still shifted more of the LP into wstETH.

## Code Snippet
[File: BLVaultLido.sol#L143-L200](https://github.com/sherlock-audit/2023-03-olympus/blob/main/sherlock-olympus/src/policies/BoostedLiquidity/BLVaultLido.sol#L143-L200)

[File: BLVaultLido.sol#L203-L256](https://github.com/sherlock-audit/2023-03-olympus/blob/main/sherlock-olympus/src/policies/BoostedLiquidity/BLVaultLido.sol#L203-L256)

## Tool used

Manual Review

## Recommendation
Consider implementing a snapshot of the entry record of OHM and wstETH and compare that with the proportionate exit record. Slash only the differential for treasury solely on dissuading large attempts to shift the pool around, and in this case it should be 0 wstETH since the originally staked wstETH is no greater than expectedWstethAmountOut.

**Comment by 0xLienid** [comment]
True, but this will be very very minor in practice. It relies on assuming no arbitrage is ever taken, it will also be helped (but not eliminated) by the solution to #003

**Comment by IAm0x52** [comment]
Sponsor has acknowledged and accepted this risk
